projects/matlib/Mat.h

```diff
@@ -0,0 +1,172 @@
+            else x = new size_t[b.ndims];
+
+            for(size_t n = 0; n < this->ndims; n++){
+                this->errorCheck(this->dims[n] != 1 && this->dims[n] != b.dims[i] && b.dims[i] != 1, "frames are not aligned");
+                if(this->dims[n] == 1) x[i] = b.dims[i];
+                else x[i] = this->dims[i];
+                i++;
+            }
+            Mat<Type> result(x[0], x[1]); //update later when higher dimension support
+            delete []x;
+
+            size_t width;
+            if(this->columns() > b.columns()) width = this->columns();
+            else width = b.columns();
+            for(size_t i = 0; i < result.size(); i++){
+                result(0,i) = this->operator()(i/width%this->rows(),i%this->columns()) + b(i/width%b.rows(),i%b.columns());
+            }
+            return result;
+        }
+        Mat operator- (Mat b){
+            size_t i = 0;
+            size_t* x;
+            if(this->ndims >= b.ndims) x = new size_t[this->ndims];
+            else x = new size_t[b.ndims];
+
+            for(size_t n = 0; n < this->ndims; n++){
+                this->errorCheck(this->dims[n] != 1 && this->dims[n] != b.dims[i] && b.dims[i] != 1, "frames are not aligned");
+                if(this->dims[n] == 1) x[i] = b.dims[i];
+                else x[i] = this->dims[i];
+                i++;
+            }
+            Mat<Type> result(x[0], x[1]); //update later when higher dimension support
+            delete []x;
+
+            size_t width;
+            if(this->columns() > b.columns()) width = this->columns();
+            else width = b.columns();
+            for(size_t i = 0; i < result.size(); i++){
+                result(0,i) = this->operator()(i/width%this->rows(),i%this->columns()) - b(i/width%b.rows(),i%b.columns());
+            }
+            return result;
+        }
+        Mat operator- (){
+            Mat<Type> result(this->rows(),this->columns());
+            for(size_t i = 0; i < this->size(); i++){
+                result(0,i) = this->operator()(0,i) * -1;
+            }
+            return result;
+        }
+        Mat operator^ (Mat b){
+            if(this->columns() != b.rows()){
+                fprintf(stderr,"Matrix size mismatch");
+                exit(1);
+            }
+            Mat<Type> result(this->rows(),b.columns());
+            Type sum;
+            for(size_t x = 0; x < this->rows(); x++){
+                for(size_t i = 0; i<b.columns();i++){
+                    sum = 0;
+                    for(size_t n = 0; n< this->columns(); n++){
+                        sum += this->operator()(x,n)*b(n,i);
+                    }
+                    result(x,i) = sum;
+                }
+            }
+            return result;
+        }
+        void T(Mat& dest){
+            this->errorCheck(this->size() != dest.size(),"Matrix size mismatch");
+            this->errorCheck(this->data == dest.data, "Source and destination matrix share same backing data");
```

Comment:
check memory, not data

Reply:
Put a TODO error to call the other transpose if same data pointer and appropriate strides+shape output

Reply:
Not sure what you mean by a TODO error. Shouldn't I just put:
`T(); return;`
?